From f95320ef8e27182d28be22109eadf2e823a26cc2 Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 09:54:19 -0400
Subject: [fs] ext4: fix online resize with a non-standard blocks per group setting

Message-id: <1414058125-4183-106-git-send-email-lczerner@redhat.com>
Patchwork-id: 98191
O-Subject: [RHEL 7.1 PATCH v2 105/171] ext4: fix online resize with a non-standard blocks per group setting
Bugzilla: 1150139
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: Theodore Ts'o <tytso@mit.edu>

BZ 1150139
https://bugzilla.redhat.com/show_bug.cgi?id=1150139

BREW 8142710
https://brewweb.devel.redhat.com/taskinfo?taskID=8142710

Upstream commit 3d2660d0c9c2f296837078c189b68a47f6b2e3b5

 The set_flexbg_block_bitmap() function assumed that the number of
 blocks in a blockgroup was sb->blocksize * 8, which is normally true,
 but not always!  Use EXT4_BLOCKS_PER_GROUP(sb) instead, to fix block
 bitmap corruption after:

 mke2fs -t ext4 -g 3072 -i 4096 /dev/vdd 1G
 mount -t ext4 /dev/vdd /vdd
 resize2fs /dev/vdd 8G

 Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
 Reported-by: Jon Bernard <jbernard@tuxion.com>
 Cc: stable@vger.kernel.org
 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/resize.c b/fs/ext4/resize.c
index 69a6261..f3b84cd 100644
--- a/fs/ext4/resize.c
+++ b/fs/ext4/resize.c
@@ -409,7 +409,7 @@ static int set_flexbg_block_bitmap(struct super_block *sb, handle_t *handle,
   start = ext4_group_first_block_no(sb, group);
   group -= flex_gd->groups[0].group;
 
-  count2 = sb->s_blocksize * 8 - (block - start);
+  count2 = EXT4_BLOCKS_PER_GROUP(sb) - (block - start);
   if (count2 > count)
    count2 = count;
 
-- 
1.7.1